iree/compiler/Conversion/HLOToHLO/DemoteF32ToF16.cpp

```diff
@@ -0,0 +1,224 @@
+ public:
+  GenericTypeConvert(MLIRContext *context, TypeConverter &converter)
+      : ConversionPattern(SourceOp::getOperationName(), 0, converter, context) {
+  }
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<NamedAttribute, 4> newAttr;
+    convertAttributes(op->getAttrs(), rewriter, newAttr);
+    llvm::SmallVector<Type, 4> newResults;
+    getTypeConverter()->convertTypes(op->getResultTypes(), newResults);
+    OperationState state(op->getLoc(), op->getName().getStringRef(), operands,
+                         newResults, newAttr, op->getSuccessors());
+    for (Region &r : op->getRegions()) {
+      Region *newRegion = state.addRegion();
+      rewriter.inlineRegionBefore(r, *newRegion, newRegion->begin());
+      TypeConverter::SignatureConversion result(newRegion->getNumArguments());
+      getTypeConverter()->convertSignatureArgs(newRegion->getArgumentTypes(),
+                                               result);
+      rewriter.applySignatureConversion(newRegion, result);
+    }
+    Operation *newOp = rewriter.createOperation(state);
+    rewriter.replaceOp(op, newOp->getResults());
+    return success();
+  }
+
+ protected:
+  static void convertAttributes(ArrayRef<NamedAttribute> attrs,
+                                ConversionPatternRewriter &rewriter,
+                                SmallVectorImpl<NamedAttribute> &newAttrs) {
+    for (auto attr : attrs) {
+      if (auto fpAttr = attr.second.dyn_cast<DenseFPElementsAttr>()) {
+        std::vector<llvm::APFloat> args;
+        if (!fpAttr.getType().getElementType().isF32()) continue;
+        for (llvm::APFloat f : fpAttr.getFloatValues()) {
+          bool losesInfo;
+          f.convert(APFloat::IEEEhalf(), APFloat::rmTowardZero, &losesInfo);
+          args.push_back(f);
+        }
+        auto tensorType = RankedTensorType::get(fpAttr.getType().getShape(),
+                                                rewriter.getF16Type());
+        newAttrs.push_back(std::make_pair(
+            attr.first, DenseElementsAttr::get(tensorType, args)));
+      } else if (auto typeAttr = attr.second.dyn_cast<TypeAttr>()) {
+        if (illegalType(typeAttr.getValue())) {
+          if (auto tensorType =
+                  typeAttr.getValue().dyn_cast<RankedTensorType>()) {
+            Type newType = RankedTensorType::get(tensorType.getShape(),
+                                                 rewriter.getF16Type());
+            newAttrs.push_back(
+                std::make_pair(attr.first, TypeAttr::get(newType)));
+          }
+        }
+      } else {
+        newAttrs.push_back(attr);
+      }
+    }
+  }
+};
+
+void ConvertF32ToF16Pass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  ModuleOp moduleOp = getOperation();
+
+  floatTypeConverter converter;
+  OwningRewritePatternList patterns;
+  patterns.insert<
```

Comment:
do you actually care about these specific ops (like, is this an allowlist)? I feel like this should just be walking all operations and converting by having a pattern that can match * (which I think you may be able to do with type conversion) - otherwise this list will be quick to get out of date (especially as we add support for TOSA and other non-mhlo things)

Reply:
I don't really care about the specific ops but I could't make the pattern matcher to work for generic ops. It seems like the pattern has to be registered for a given op name. Do you know if there is a way to make the pattern apply to all ops? Otherwise I can give up on the pattern matcher but I don't think I can use the type converter in that case. Any thoughts?

Reply:
Look for MatchAnyOpTypeTag for how to not match on a specific op name.

See here for an example: https://github.com/llvm/llvm-project/blob/ed95cafbc5fa9efbfe3f38da0b17efdb3806598c/mlir/lib/Dialect/Linalg/Transforms/ElementwiseToLinalg.cpp#L29

Reply:
Thanks! I was looking for it for some time... Changed it to use MatchAnyOpTypeTag.